packages/vite/src/node/plugins/importMetaGlob.ts

```diff
@@ -51,7 +51,15 @@ export function getAffectedGlobModules(
 ): ModuleNode[] {
   const modules: ModuleNode[] = []
   for (const [id, allGlobs] of server._importGlobMap!) {
-    if (allGlobs.some((glob) => isMatch(file, glob)))
+    // (glob1 || glob2) && !glob3 && !glob4...
+    if (
+      allGlobs.some((glob) =>
+        glob.reduce((match, g) => {
+          const m = isMatch(file, g)
+          return g[0] === '!' ? match && m : match || m
+        }, glob[0] === '!'),
+      )
+    )
```

Comment:
```suggestion
    if (
      allGlobs.reduce((acc, glob) =>
        const m = isMatch(file, glob)
        return glob[0] === '!' ? match && m : match || m
      }, allGlobs[0][0] === '!')
    )
```
Shouldn't this be like this? `glob.reduce` will iterate over all characters of the glob in your code.

Though, I think `.reduce` is hard to understand and would be better to write it like:
```js
const { affirmed, negated } = groupGlobs(allGlobs)
if (
  affirmed.some(glob => isMatch(file, glob)) &&
  negated.every(glob => !isMatch(file, glob))
)
```

Reply:
### About option1:

```js
if (
      allGlobs.reduce((acc, glob) =>
        const m = isMatch(file, glob) 
        return glob[0] === '!' ? match && m : match || m
      }, allGlobs[0][0] === '!')
    )
```

`isMatch` basically use the `OR` operation, but we need `AND` operation here as https://github.com/vitejs/vite/issues/13374#issuecomment-1607230035 said

### About option2:

```js
const { affirmed, negated } = groupGlobs(allGlobs)
if (
  affirmed.some(glob => isMatch(file, glob)) &&
  negated.every(glob => !isMatch(file, glob))
)
```

Consider `allGlobs` below:

```js
[
  [g1, neg1],  // rule1
  [g2, neg2], // rule2
]
```
Glob paths dont have to match all the `negated` patterns, in fact they just need to match either `rule1`
 or `rule2`.

Reply:
> * Isn't `glob`'s type `string[]` ?

Oh, my bad🤦Please forgot about the first one.

For the second one I meant:
```js
if (
  allGlobs.some(globs => {
    const { affirmed, negated } = groupGlobs(globs)
    return (
      affirmed.some(glob => isMatch(file, glob)) &&
      negated.every(glob => isMatch(file, glob))
    )
  })
)
```

Maybe it's better to set a function when setting `server._importGlobMap`:
```js
server._importGlobMap.set(id, allGlobs.map(globs => {
  const { affirmed, negated } = groupGlobs(globs)
  return file => (
    affirmed.some(glob => isMatch(file, glob)) &&
    negated.every(glob => isMatch(file, glob))
  )
}))
```

Reply:
My initial idea was to reduce the iterations of  `globs`, but I'm ok with `const { affirmed, negated } = groupGlobs(globs)`.

> Maybe it's better to set a function when setting server._importGlobMap:

Not sure whether we should change the `_importGlobMap` values, I think we could leave it `string[][]`  for now

Reply:
Yeah, we can achieve it in a single loop by using `.reduce`.
`groupGlobs` will require two loops. But if we change the `_importGlobMap` values and precompute that value, we can reduce it to a single loop. (The precomputation will take a single loop but it can be ignored as it will only run once.)

Reply:
I changed `_importGlobMap` values' type to `{ affirmed: string[]; negated: string[] }[]`, it would be more flexible.